### ISSUE TITLE
Generic test method with type argument in Func throws

### ISSUE THREAD

**Issue by ViktorHofer** [xunit/xunit#1794]
Repro:

```csharp
[Theory]
[MemberData(nameof(CreateRangeSliceWithSelectorData))]
public void CreateRangeSliceWithSelector<TResult>(IEnumerable<int> source, int start, int length, Func<int, TResult> selector)
{
    var expected = source.Skip(start).Take(length).Select(selector);
    Assert.Equal(expected, ImmutableArray.CreateRange(source.ToImmutableArray(), start, length, selector));
}

public static IEnumerable<object[]> CreateRangeSliceWithSelectorData()
{
    yield return new object[] { new[] { 4, 5, 6, 7 }, 0, 0, new Func<int, float>(i => i + 0.5f) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 0, 0, new Func<int, double>(i => i + 0.5d) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 0, 0, new Func<int, int>(i => i) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 0, 1, new Func<int, int>(i => i * 2) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 0, 2, new Func<int, int>(i => i + 1) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 0, 4, new Func<int, int>(i => i) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 3, 1, new Func<int, int>(i => i) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 3, 0, new Func<int, int>(i => i) };
    yield return new object[] { new[] { 4, 5, 6, 7 }, 4, 0, new Func<int, int>(i => i) };
}
```

Exception:

```
System.Collections.Immutable.Tests.ImmutableArrayTest.CreateRangeSliceWithSelector [FAIL]
        System.IndexOutOfRangeException : Index was outside the bounds of the array.
```

This isn't really blocking us in corefx as we just disable these tests for now (only 4 tests). If I have time I will look into it.

**Comment by ViktorHofer** [comment]
cc @hughbe

**Comment by bradwilson** [comment]
Is this related to #1781? Or something different?

**Comment by ViktorHofer** [comment]
It's a different one. I believe it's related to https://github.com/xunit/xunit/pull/965. Probably an oversight.

**Comment by bradwilson** [comment]
Fixed by #1825.
